From e9d2dd3922d8b9fb1efe2153c08705afdaff518c Mon Sep 17 00:00:00 2001
From: Harini Katakam <harini.katakam@xilinx.com>
Date: Thu, 24 Sep 2020 13:16:49 +0530
Subject: [PATCH 1611/1851] net: xilinx: XXV Ethernet HW timestamp fixes

commit d42de16ed2832785de4066fac3883d46ed3f063f from
https://github.com/Xilinx/linux-xlnx.git

When mcdma error occurs, the BDs are reset. Correspondingly, the
timestamp FIFOs should also be reset.

In timestamp_mode function, there's an unnecessary read of a register
offset that is not defined or relevant for XXV. Remove the same.

Signed-off-by: Harini Katakam <harini.katakam@xilinx.com>
Reviewed-by: Radhey Shyam Pandey <radhey.shyam.pandey@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet_main.c  |  6 ++++--
 drivers/net/ethernet/xilinx/xilinx_axienet_mcdma.c | 13 +++++++++++++
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index f9e4336b1854..0f90aaf13b05 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -1954,7 +1954,8 @@ static int axienet_set_timestamp_mode(struct axienet_local *lp,
 		return -EINVAL;
 
 	/* Read the current value in the MAC TX CTRL register */
-	regval = axienet_ior(lp, XAE_TC_OFFSET);
+	if (lp->axienet_config->mactype != XAXIENET_10G_25G)
+		regval = axienet_ior(lp, XAE_TC_OFFSET);
 
 	switch (config->tx_type) {
 	case HWTSTAMP_TX_OFF:
@@ -1976,7 +1977,8 @@ static int axienet_set_timestamp_mode(struct axienet_local *lp,
 		axienet_iow(lp, XAE_TC_OFFSET, regval);
 
 	/* Read the current value in the MAC RX RCW1 register */
-	regval = axienet_ior(lp, XAE_RCW1_OFFSET);
+	if (lp->axienet_config->mactype != XAXIENET_10G_25G)
+		regval = axienet_ior(lp, XAE_RCW1_OFFSET);
 
 	/* On RX always timestamp everything */
 	switch (config->rx_filter) {
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_mcdma.c b/drivers/net/ethernet/xilinx/xilinx_axienet_mcdma.c
index 6c18e51ed515..5c13fcd1011e 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_mcdma.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_mcdma.c
@@ -723,6 +723,19 @@ void __maybe_unused axienet_mcdma_err_handler(unsigned long data)
 	if (lp->axienet_config->mactype != XAXIENET_10G_25G)
 		axienet_iow(lp, XAE_FCC_OFFSET, XAE_FCC_FCRX_MASK);
 
+#ifdef CONFIG_XILINX_AXI_EMAC_HWTSTAMP
+	if (lp->axienet_config->mactype == XAXIENET_10G_25G) {
+		axienet_rxts_iow(lp, XAXIFIFO_TXTS_RDFR,
+				 XAXIFIFO_TXTS_RESET_MASK);
+		axienet_rxts_iow(lp, XAXIFIFO_TXTS_SRR,
+				 XAXIFIFO_TXTS_RESET_MASK);
+		axienet_txts_iow(lp, XAXIFIFO_TXTS_RDFR,
+				 XAXIFIFO_TXTS_RESET_MASK);
+		axienet_txts_iow(lp, XAXIFIFO_TXTS_SRR,
+				 XAXIFIFO_TXTS_RESET_MASK);
+	}
+#endif
+
 	lp->axienet_config->setoptions(ndev, lp->options &
 				       ~(XAE_OPTION_TXEN | XAE_OPTION_RXEN));
 	axienet_set_mac_address(ndev, NULL);
-- 
2.31.1

